src/library/scala/collection/immutable/HashSet.scala

```diff
@@ -1013,6 +1017,271 @@ private final class BitmapIndexedSetNode[A](
+      var anyChangesMadeSoFar = false
+
+      val allMap = dataMap | bm.dataMap | nodeMap | bm.nodeMap
+
+      // minimumIndex is inclusive -- it is the first index for which there is data or nodes
+      val minimumBitPos: Int = Node.bitposFrom(Integer.numberOfTrailingZeros(allMap))
+      // maximumIndex is inclusive -- it is the last index for which there is data or nodes
+      // it could not be exclusive, because then upper bound in worst case (Node.BranchingFactor) would be out-of-bound
+      // of int bitposition representation
+      val maximumBitPos: Int = Node.bitposFrom(Node.BranchingFactor - Integer.numberOfLeadingZeros(allMap) - 1)
+
+      var leftNodeRightNode = 0
+      var leftDataRightNode = 0
+      var leftNodeRightData = 0
+      var leftDataOnly = 0
+      var rightDataOnly = 0
+      var leftNodeOnly = 0
+      var rightNodeOnly = 0
+      var leftDataRightDataMigrateToNode = 0
+      var leftDataRightDataLeftOverwrites = 0
+
+      var dataToNodeMigrationTargets = 0
+
+    {
+      var bitpos = minimumBitPos
+      var leftIdx = 0
+      var rightIdx = 0
+      var finished = false
+
+      while (!finished) {
+
+        if ((bitpos & dataMap) != 0) {
+          if ((bitpos & bm.dataMap) != 0) {
+            if (getHash(leftIdx) == bm.getHash(rightIdx) && getPayload(leftIdx) == bm.getPayload(rightIdx)) {
+              leftDataRightDataLeftOverwrites |= bitpos
+            } else {
+              leftDataRightDataMigrateToNode |= bitpos
+              dataToNodeMigrationTargets |= Node.bitposFrom(Node.maskFrom(improve(getHash(leftIdx)), shift))
+            }
+            rightIdx += 1
+          } else if ((bitpos & bm.nodeMap) != 0) {
+            leftDataRightNode |= bitpos
+          } else {
+            leftDataOnly |= bitpos
+          }
+          leftIdx += 1
+        } else if ((bitpos & nodeMap) != 0) {
+          if ((bitpos & bm.dataMap) != 0) {
+            leftNodeRightData |= bitpos
+            rightIdx += 1
+          } else if ((bitpos & bm.nodeMap) != 0) {
+            leftNodeRightNode |= bitpos
+          } else {
+            leftNodeOnly |= bitpos
+          }
+        } else if ((bitpos & bm.dataMap) != 0) {
+          rightDataOnly |= bitpos
+          rightIdx += 1
+        } else if ((bitpos & bm.nodeMap) != 0) {
+          rightNodeOnly |= bitpos
+        }
+
+        if (bitpos == maximumBitPos) {
+          finished = true
+        } else {
+          bitpos = bitpos << 1
+        }
+      }
+    }
+
+
+      val newDataMap = leftDataOnly | rightDataOnly | leftDataRightDataLeftOverwrites
+
+      val newNodeMap =
+        leftNodeRightNode |
+          leftDataRightNode |
+          leftNodeRightData |
+          leftNodeOnly |
+          rightNodeOnly |
+          dataToNodeMigrationTargets
+
+
+      if ((newDataMap == (leftDataOnly | leftDataRightDataLeftOverwrites)) && (newNodeMap == leftNodeOnly)) {
+        // nothing from `this` will make it into the result -- return early
```

Comment:
comment should read "nothing from `bm` will make it into the result -- return early"